p4-16/spec/P4-16-spec.mdk

```diff
@@ -7097,16 +7100,155 @@ annotationBody
     | annotationBody annotationToken
 ~ End P4Grammar
 
-Annotations may impose additional structure on their bodies, and are not
-confined to the P4 language. For example, the P4Runtime specification defines a
-`@pkginfo` annotation that takes key-value pairs, which are not found in the
-main P4 language.
+Unstructured annotations may impose additional structure on their bodies, and
+are not confined to the P4 language. For example, the P4Runtime specification
+defines a `@pkginfo` annotation that takes key-value pairs, which are not found
+in the main P4 language.
+
+## Bodies of Structured Annotations
+
+Unlike unstructured annotations, structured annotations use square brackets
+`[...]` and have a restricted format. They are commonly used to declare custom
+metadata, consisting of expression lists or key-value lists but not both. An
+`expressionList` may be empty or contain a comma-separated list of member
+`expression`s. A `kvList` consists of one or more `kvPair`s, each consisting of
+a key and a value `expression`. Note the syntax for `expression` is rich, see
+Appendix [#sec-grammar] for details.
+
+To avoid ambiguity, an empty `stucturedAnnotationBody` is by definition
+considered to contain an empty `expressionList`.
+
+All `expression`s within a `stucturedAnnotationBody` must evaluate at
+compile-time to one of the following base types: string literals,
+infinite-precision integers or booleans. Nested `expression`s (e.g. an
+`expression` containing an `expressionList`, a `kvList`, etc.) are not
+allowed. Note that P4Runtime information (`P4Info`) may stipulate additional
+restrictions. For example, an integer expression might be limited to 64-bit
+values.
+
+It is illegal to duplicate a key within an annotation.  It is illegal to use the
+same annotation `name` on a given `annotationToken`.
```

Comment:
"... same annotation `name` on a given `annotationToken` for structured annotations". Since I assume this is only a restriction we have for structured annotations.